src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/util/Utils.java

```diff
@@ -17,10 +47,62 @@ public static void configureMetrics() {
         MetricsConfiguration.CONFIG_MAP.put(NodeDetailsCollector.class, cdefault);
-        MetricsConfiguration.CONFIG_MAP.put(NodeStatsMetricsCollector.class, cdefault);
+        MetricsConfiguration.CONFIG_MAP.put(NodeStatsAllShardsMetricsCollector.class, cdefault);
+        MetricsConfiguration.CONFIG_MAP.put(NodeStatsFixedShardsMetricsCollector.class, cdefault);
         MetricsConfiguration.CONFIG_MAP.put(MasterServiceEventMetrics.class, new MetricsConfiguration.MetricConfig(1000, 0, 0));
         MetricsConfiguration.CONFIG_MAP.put(MasterServiceMetrics.class, cdefault);
         
     }
 
+    // These methods are utility functions for the Node Stat Metrics Collectors. These methods are used by both the all
+    // shards collector and the few shards collector.
+
+    /**
+     * This function is copied directly from IndicesService.java in elastic search as the original function is not public
+     * we need to collect stats per shard based instead of calling the stat() function to fetch all at once(which increases
+     * cpu usage on data nodes dramatically).
+     * @param indicesService Indices Services which keeps tracks of the indexes on the node
+     * @param indexShard Shard to fetch the metrics for
+     * @param flags  The Metrics Buckets which needs to be fetched.
+     * @return stats given in the flags param for the shard given in the indexShard param.
+     */
+    public static IndexShardStats indexShardStats(final IndicesService indicesService, final IndexShard indexShard,
+                                                  final CommonStatsFlags flags) {
+        if (indexShard.routingEntry() == null) {
+            return null;
+        }
+
+        return new IndexShardStats(
+                indexShard.shardId(),
+                new ShardStats[]{
+                        new ShardStats(
+                                indexShard.routingEntry(),
+                                indexShard.shardPath(),
+                                new CommonStats(indicesService.getIndicesQueryCache(), indexShard, flags),
+                                null,
+                                null,
+                                null)
+                });
+    }
+
+    public static HashMap<String, IndexShard> getShards() {
+        HashMap<String, IndexShard> shards =  new HashMap<>();
+        Iterator<IndexService> indexServices = ESResources.INSTANCE.getIndicesService().iterator();
+        while (indexServices.hasNext()) {
+            Iterator<IndexShard> indexShards = indexServices.next().iterator();
+            while (indexShards.hasNext()) {
+                IndexShard shard = indexShards.next();
+                shards.put(getUniqueShardIdKey(shard.shardId()), shard);
+            }
+        }
+        return shards;
+    }
+
+    public static String getUniqueShardIdKey(ShardId shardId) {
+        return "[" + shardId.hashCode() + "][" + shardId.getId() + "]";
```

Comment:
I think this will be same for both pri and rep shards. You will have to add that info here if you want them to be unique.

Reply:
> I think this will be same for both pri and rep shards. You will have to add that info here if you want them to be unique.

ShardId's hashcode is generated with UUID. Is it sufficient to differentiate among primary/replica ?

Reply:
> ShardId's hashcode is generated with its UUID. Is it sufficient to differentiate among primary/replica ?

No, the hashcode just uses `index.hashcode` and the `int` shardId. In general, in ES, there is no difference maintained b/w pri and replica shards. We generally require `shardrouting.primary()` if there is a need to differentiate between the two. We first need to figure out if we need to differentiate b/w pri and rep here. That depends on the target consumers of this util fn.

Reply:
The node stats metrics don't have a ShardRole as a dimension. I am not sure if that was intentional. They just have a [ShardID and IndexName](https://opendistro.github.io/for-elasticsearch-docs/docs/pa/reference/) as the dimensions. Currently we would only emit these metrics for the shards present on a node. Since no node would have the primary and replica of the same shard on the same node, there is no conflict and metric is emitted for both the primary and the replica.